minetest/minetest
# Render shadows on entities.

Fixes a number of problems with how shadows are applied to entities:
* Bug with mod 'drawers' described in #11326
* Colored shadows not being applied properly to entities
* Different shader logic being used for entities and nodes
* Light and color of entities being calculated and communicated differently from nodes
* Entity surface not being shaded correctly when facing way from directional light

What I've done so far:
* I've added code to assign SM texture to its texture layer when setting texture in content_cao.
* Implemented the same light color contract for entities as for nodes (alpha communicating the balance between natural/artificial). The only difference is that emissiveColor is used as a factor to the vertex color.
* Copied the SM logic from node shader and object shader.
* Fixed the way colored shadows are applied at small falling angle (cos(alpha) < 0.035 for nodes and 0.14 for objects)

What's left
- [x] Fix the way split-normals are loaded, since it causes problems with shading (e.g. MCL2 chests, boats, wolves)
- [x] Calibrate offset bias to remove self-shadowing on dropped node stacks (vanilla Minetest)
- [x] Fix phantom shadows from entities (looks like a broken world transformation)
- [x] (Almost) Fix self-shadowing of banner entities in MCL2 (offsets?)

This PR is Ready for Review

## How to test

* Add 'drawers' mod
* Place three drawers: one in a shadow, one on the edge and one in the sunlight.
* Place node of sand into each of the drawers (right-click)
* Drawers should display sand inventory image, properly shaded together with the drawer node itself.

Another test:
* Place any entity (e.g. animals, chests and boats in MCL2).
* Fly/walk over.
* Observe your shadow rendered on the entity.
